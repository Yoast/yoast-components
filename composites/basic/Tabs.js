import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

const Tab = styled.section`
	padding: 18px 20px;
	display: none;
	
	&.tab-active {
		display: block;
	}
`;

/**
 * Creates a series of tabs.
 */
class Tabs extends React.Component {
	/**
	 * Constructs the class and sets its initial state.
	 *
	 * @param {Object} props The props to use within the component.
	 *
	 * @constructor
	 */
	constructor( props ) {
		super( props );

		this.state = { activeTab: this.props.activeTab };
		this.getTabs();
	}

	/**
	 * Checks whether props were changed and applies them to the state to change the currently active tab.
	 *
	 * @param {Object} newProps The new props to use.
	 *
	 * @returns {void}
	 */
	componentWillReceiveProps( newProps ) {
		this.setState( { activeTab: newProps.activeTab } );
	}

	/**
	 * Gets all the defined tabs and returns an array of ReactElements.
	 *
	 * @returns {Array} Array containing a Tab component for every defined item found in the props.
	 */
	getTabs() {
		return this.props.items.map( ( item ) => {
			return this.addTab( item.id, item.url, item.view );
		} );
	}

	/**
	 * Adds the Tab component.
	 *
	 * @param {string} id The unique ID for the tab.
	 * @param {string} url The current URL to check against.
	 * @param {ReactElement} content The content of the tab.
	 *
	 * @returns {ReactElement} The rendered Tab component.
	 */
	addTab( id, url, content ) {
		return <Tab key={ id } className={ url === this.state.activeTab ? "tab-active" : "" }>
			{ content }
		</Tab>;
	}

	/**
	 * Renders out all the tabs.
	 *
	 * @returns {ReactElement} The tabs.
	 */
	render() {
		return <div>
			{ this.getTabs() }
		</div>;
	}
}

Tabs.propTypes = {
	items: PropTypes.Array,
	isActive: PropTypes.bool,
	activeTab: PropTypes.string.isRequired,
};

Tabs.defaultProps = {
	items: [],
	isActive: false,
};

export default Tabs;
