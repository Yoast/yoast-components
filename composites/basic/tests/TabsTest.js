import React from "react";
import renderer from "react-test-renderer";

import Tabs from "../Tabs.js";

let items = [
	{ id: "tab1", url: "http://example.com/tab1", view: <p>This is some content</p> },
	{ id: "tab2", url: "http://example.com/tab2", view: <p>This is some content</p> },
	{ id: "tab3", url: "http://example.com/tab3", view: <p>This is some content</p> },
];

test( "the SubNavigation with active first tab matches the snapshot", () => {
	const component = renderer.create(
	<Tabs items={ items } activeTab="http://example.com/tab1"/>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
