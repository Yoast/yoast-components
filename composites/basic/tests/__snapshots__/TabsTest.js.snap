// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SubNavigation with active first tab matches the snapshot 1`] = `
<div>
  <section
    className="tab-active sc-bdVaJa hboHnN"
  >
    <p>
      This is some content
    </p>
  </section>
  <section
    className="sc-bdVaJa hboHnN"
  >
    <p>
      This is some content
    </p>
  </section>
  <section
    className="sc-bdVaJa hboHnN"
  >
    <p>
      This is some content
    </p>
  </section>
</div>
`;
